Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,421 @@
+    /**
+     * @var string
+     */
+    private $savePath;
+    /**
+     * @var string
+     */
+    private $sessionName;
+    /**
+     * @var Transaction
+     */
+    private $transaction;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param string $projectId The current project id.
+     * @param string $database The database id.
+     * @param array $options [optional] {
+     *     Configuration Options.
+     *
+     *     @type int $gcLimit The number of entities to delete in the garbage
+     *        collection. Values larger than 1000 will be limited to 1000.
+     *        **Defaults to** `0`, indicating garbage collection is disabled by
+     *        default.
+     *     @type string $collectionNameTemplate A sprintf compatible template
+     *        for formatting the collection name where sessions will be stored.
+     *        The template receives two values, the first being the save path
+     *        and the latter being the session name.
+     *     @type array $begin Configuration options for beginTransaction.
+     *     @type array $commit Configuration options for commit.
+     *     @type array $rollback Configuration options for rollback.
+     *     @type array $read Configuration options for read.
+     *     @type array $query Configuration options for runQuery.
+     * }
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        $projectId,
+        $database,
+        array $options = []
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->projectId = $projectId;
+        $this->database = $database;
+        $this->options = $options + [
+            'begin' => [],
+            'commit' => [],
+            'rollback' => [],
+            'query' => [],
+            'read' => [],
+            'gcLimit' => 0,
+            'collectionNameTemplate' => '%1$s:%2$s',
+        ];
+
+        // Cut down gcLimit to 1000
+        $this->options['gcLimit'] = min($this->options['gcLimit'], 1000);
+    }
+
+    /**
+     * Start a session, by getting the Firebase collection from $savePath.
+     *
+     * @param string $savePath The value of `session.save_path` setting will be
+     *        used here as the Firestore collection ID.
+     * @param string $sessionName The value of `session.name` setting will be
+     *        used here as the Firestore document ID prefix (ex: "PHPSESSID:").
+     * @return bool
+     */
+    public function open($savePath, $sessionName)
+    {
+        $this->savePath = $savePath;
+        $this->sessionName = $sessionName;
+        $database = $this->databaseName($this->projectId, $this->database);
+
+        $beginTransaction = $this->connection->beginTransaction([
```

Comment:
Should we wrap this in a try/catch block as well?

Reply:
If we do this, `open` will return `false` instead of throwing an exception if it fails. This seems consistent and correct to me.